### PR TITLE
fix(graphql, rest): allow caching of options requests

### PR DESCRIPTION
**Pull request by damienwebdev**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/graycoreio/magento2-cors/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?
Currently, Varnish will incorrectly cache the result of CORS requests.

Fixes: #52 

## What is the new behavior?
Varnish will now appropriate cache CORS requests per origin.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

### CODE REVIEW

File: Response/HeaderProvider/CorsVaryHeaderProvider.php [hunk 1]
```diff
@@ -63,9 +50,12 @@ public function __construct(
      * ```
      *
      * So, we don't bother attempting to merge the two hears into a comma-separated list.
```
Comment (griest024) [inline Response/HeaderProvider/CorsVaryHeaderProvider.php:52]: `hears` -> `headers`

File: Response/HeaderProvider/CorsVaryHeaderProvider.php [hunk 2]
```diff
@@ -63,9 +50,12 @@ public function __construct(
...
+     *
+     * //TODO(damienwebdev): consider the ramifications to cache size if we don't
+     * normalize "Vary" this into well organized list.
      */
     public function canApply(): bool
     {
-        return $this->validator->originIsValid() && $this->getValue();
+        return true;
```
Comment (griest024) [inline Response/HeaderProvider/CorsVaryHeaderProvider.php:59]: Where does the origin validation happen? Is this behavior that should be baked intot he module or is it just a special case to handle origins differently?

**Review by griest024 — CHANGES_REQUESTED** [review]

File: Response/Preflight/GraphQl/PreflightRequestHandler.php [hunk 3]
```diff
@@ -47,7 +47,7 @@ public function aroundDispatch(GraphQlController $subject, callable $next, Reque
     {
         if ($request instanceof Http && $request->isOptions()) {
             $this->_headerManager->beforeSendResponse($this->_response);
-            $this->_response->setNoCacheHeaders();
+            $this->_response->setPublicHeaders(86400);
```
Comment (griest024) [inline Response/Preflight/GraphQl/PreflightRequestHandler.php:50]: `setPublicHeaders` is setting the cache max age I assume?

Comment (damienwebdev) [reply Response/Preflight/GraphQl/PreflightRequestHandler.php:50]: Correct, it sets the cache age for Varnish, not the browser cache. Browser only respects `Access-control-max-age` whereas varnish determines its cache from the Cache-control max age.

File: Response/HeaderProvider/CorsVaryHeaderProvider.php (hunk 2 above)
Comment (damienwebdev) [reply Response/HeaderProvider/CorsVaryHeaderProvider.php:59]: We don't need to validate Origin for the `vary` header at all. We're simply telling all HTTP client's that this header will vary, so use it as a cache key for cache lookups.

File: Response/HeaderProvider/CorsVaryHeaderProvider.php (hunk 1 above)
Comment (damienwebdev) [reply Response/HeaderProvider/CorsVaryHeaderProvider.php]: 2f578696c5ac98282252eab14f7c00e3db186581

**Review by griest024 — APPROVED** [review]

**Pull request merged**